Lib/test/test_asyncio/test_locks.py

```diff
@@ -903,10 +907,32 @@ async def test_acquire_hang(self):
 
         t1.cancel()
         sem.release()
-        await asyncio.sleep(0)
+        await asyncio.sleep(0.01)
         self.assertTrue(sem.locked())
         self.assertTrue(t2.done())
 
+    async def test_acquire_no_hang(self):
+
+        sem = asyncio.Semaphore(1)
+
+        async def c1():
+            async with sem:
+                await asyncio.sleep(0)
+            t2.cancel()
+
+        async def c2():
+            async with sem:
+                self.assertFalse(True)
+
+        t1 = asyncio.create_task(c1())
+        t2 = asyncio.create_task(c2())
+
+        result = await asyncio.gather(t1, t2, return_exceptions=True)
+        self.assertTrue(result[0] is None)
+        self.assertTrue(isinstance(result[1], asyncio.CancelledError))
+
+        await asyncio.wait_for(sem.acquire(), timeout=0.01)
```

Comment:
For some reason this timeout still needs to be 0.01. I'll look into it.

Reply:
This is because of how `wait_for()` is implemented -- using `print()` I verified that at this point the semaphore is unlocked and its value is 1, and the `acquire()` call takes a fraction of a millisecond. But `wait_for()` requires two iterations through the event loop -- the first to call `_release_waiter()`, then the second to wake up `wait_for()`. I'd specify a large timeout (e.g. 1.0) so hiccups in CI are less likely to cause test flakes -- we now know that the test passes, and if it were to fail it would just hang forever.

Reply:
Because this test doesn't sleep -- if the test passes the timeout isn't reached. My explanation was not very clear -- the problem is not that `wait_for()` requires two event loop iterations (though it does), but that the timeout means something completely different than the sleep time in other places.